source/extensions/filters/http/ext_proc/mutation_utils.cc

```diff
@@ -44,10 +44,72 @@ void MutationUtils::headersToProto(const Http::HeaderMap& headers_in,
   });
 }
 
-absl::Status MutationUtils::applyHeaderMutations(const HeaderMutation& mutation,
+absl::Status MutationUtils::responseHeaderSizeCheck(const uint32_t max_request_headers_kb,
+                                                    const uint32_t max_request_headers_count,
+                                                    const HeaderMutation& mutation,
+                                                    Counter& rejected_mutations) {
+  const auto& remove_headers = mutation.remove_headers();
+  if (remove_headers.size() > static_cast<int>(max_request_headers_count) ||
+      remove_headers.SpaceUsedExcludingSelf() > static_cast<int>(max_request_headers_kb * 1024)) {
+    ENVOY_LOG(debug,
+              "Mutation remove header count {} or mutation remove header size {} may exceed the "
+              "limit. Returning error.",
+              remove_headers.size(), remove_headers.SpaceUsedExcludingSelf());
+    rejected_mutations.inc();
+    return absl::InvalidArgumentError(absl::StrCat(
+        "Remove header count ", std::to_string(remove_headers.size()), " or remove header size ",
+        std::to_string(remove_headers.SpaceUsedExcludingSelf()),
+        " exceed the HCM header size limit."));
+  }
+
+  const auto& set_headers = mutation.set_headers();
+  if (set_headers.size() > static_cast<int>(max_request_headers_count) ||
+      set_headers.SpaceUsedExcludingSelf() > static_cast<int>(max_request_headers_kb * 1024)) {
```

Comment:
We should add the counter check before the for loop.  But we don't add the size check outside of the for loop.

The size check should be done inside the for loop when working on each header, and reject if the size over the limits.